validation/validate.py

```diff
@@ -0,0 +1,267 @@
+
+async def evaluate_single_result(
+    evaluator: QueryEvaluator,
+    result: Dict,
+    groundtruth: Dict,
+    sys_message: str,
+) -> pd.DataFrame:
+    """Evaluate a single result"""
+    try:
+        eval_result = await evaluator.get_eval_results(
+            groundtruth_data=[groundtruth],
+            responses_to_evaluate=[result],
+            instructions=sys_message,
+        )
+        eval_result["db_name"] = result["db_name"]
+        return eval_result
+    except Exception as e:
+        print(f"Error evaluating result: {e}")
+        return pd.DataFrame(
+            [
+                {
+                    "db_name": result["db_name"],
+                    "question": groundtruth["question"],
+                    "question_metadata": groundtruth["question_metadata"],
+                    "eval_status": f"Error: {str(e)}",
+                    "request_status": result["request_status"],
+                }
+            ]
+        )
+
+
+async def evaluate_results(
+    evaluator: QueryEvaluator,
+    llm_responses: pd.DataFrame,
+    test_cases: pd.DataFrame,
+    sys_message: str,
+) -> pd.DataFrame:
+    groundtruth_data = reformat_input_data(test_cases)
+    responses_to_evaluate = llm_responses.to_dict(orient="records")
+
+    evaluation_tasks = [
+        evaluate_single_result(
+            evaluator=evaluator,
+            result=result,
+            groundtruth=groundtruth_data[i],
+            sys_message=sys_message,
+        )
+        for i, result in enumerate(responses_to_evaluate)
+    ]
+
+    all_eval_results = await asyncio.gather(*evaluation_tasks)
+
+    return pd.concat(all_eval_results, ignore_index=True)
+
+
+def _convert_columns(df: pd.DataFrame, column_name: str, type: str) -> pd.DataFrame:
+    if type == "list":
+        df[column_name] = df[column_name].apply(lambda x: ast.literal_eval(x))
+    elif type == "dict":
+        df[column_name] = df[column_name].apply(lambda x: x.replace("'", '"'))
+        df[column_name] = df[column_name].apply(lambda x: json.loads(x))
+    return df
+
+
+def reformat_input_data(input_data: pd.DataFrame) -> List[Dict]:
+    conversion_dict = {
+        "correct_best_tables": "list",
+        "tests_to_run": "list",
+        "correct_best_columns": "dict",
+        "question_metadata": "dict",
+    }
+
+    for column_name, type in conversion_dict.items():
+        input_data = _convert_columns(input_data, column_name, type)
+
+    return [row.to_dict() for _, row in input_data.iterrows()]
+
+
+def get_async_session(db_name: str):
+    aengine = create_async_engine(
+        url=f"sqlite+aiosqlite:///{DATA_SOURCES_PATH}/{db_name}.sqlite",
+    )
+    async_session = sessionmaker(
+        bind=aengine, class_=AsyncSession, expire_on_commit=False
+    )
+    return async_session
+
+
+if __name__ == "__main__":
+    time_start = time.time()
+    asyncio.run(main())
+    # Rounded to 2 decimal places
+    print(f"Total time taken: {round(time.time() - time_start, 2)}s")
```

Comment:
```suggestion
if __name__ == "__main__":
    import argparse
    parser = argparse.ArgumentParser()
    parser.add_argument("--path_to_data_sources", type=str, default=None)
    
    # Add other arguments if you want

    time_start = time.time()

    args = parser.parse_args()
    asyncio.run(main(args))

    # Rounded to 2 decimal places
    print(f"Total time taken: {round(time.time() - time_start, 2)}s")
```
Then you can run from cmdline with `python validate.py --path_to_data_sources="../../demo_databases/" --<some other arguments>`